chips/nrf52840/src/ieee802154_radio.rs

```diff
@@ -780,87 +803,109 @@ impl<'a> Radio<'a> {
+                // exhibit undefined behavior.
+                ////////////////////////////////////////////////////////////////
 
                 // Since READY_START shortcut enabled, always clear READY event
                 self.registers.event_ready.write(Event::READY::CLEAR);
 
                 // Completed receiving a packet, now determine if we need to send ACK
                 if self.registers.event_end.is_set(Event::READY) {
                     self.registers.event_end.write(Event::READY::CLEAR);
-                    let result = self.crc_check();
+                    let crc = self.crc_check();
 
-                    // Unwrap fail = Radio RX Buffer is missing (may be due to receive client not replacing in receive(...) method,
-                    // or some instance in  driver taking buffer without properly replacing).
+                    // Unwrap fail = Radio RX Buffer is missing (may be due to
+                    // receive client not replacing in receive(...) method, or
+                    // some instance in driver taking buffer without properly
+                    // replacing).
                     let rbuf = self.rx_buf.take().unwrap();
 
-                    // data buffer format: | PSDU OFFSET | FRAME | LQI |
-                    // length of received data transferred to buffer (including PSDU)
-                    let data_len = rbuf[MIMIC_PSDU_OFFSET as usize] as usize;
-
-                    // Check if the received packet has a valid CRC and as a last resort
-                    // confirm that the received packet length field is in compliance
-                    // with the maximum packet length. If not, drop the packet.
-                    if !result.is_ok() || data_len >= MAX_FRAME_SIZE + PSDU_OFFSET {
-                        self.rx_buf.replace(rbuf);
-                        return
-                    }
-
-                    // lqi is found just after the data received
+                    // Data buffer format: | PREFIX | PHR | PSDU | LQI |
+                    //
+                    // Retrieve the length of the PSDU (actual frame). The frame
+                    // length is only 7 bits, but of course the field is a byte.
+                    // The nRF52840 product specification says this about the
+                    // PHR byte (Version 1.8, section 6.20.12.1):
+                    //
+                    // > The most significant bit is reserved and is set to zero
+                    // > for frames that are standard compliant. The radio
+                    // > module will report all eight bits and it can
+                    // > potentially be used to carry some information.
+                    //
+                    // We are not using that for any information so we just
+                    // force it to zero. This ensures that `data_len` will not
+                    // be longer than our buffer.
+                    let data_len = (rbuf[radio::PHR_OFFSET] & 0x7F) as usize;
```

Comment:
Since we do not immediately drop the packet on a failed CRC, this indexing will still occur on corrupted packets and may lead to an indexing error.

Reply:
How? How can `rx_buf` be less than 2 bytes long? I mean I guess it could be, but the 15.4 stack is not going to work.

If you mean later with `data_len`, since that now maxes at 127, which is smaller than the required buffer size, the same logic should apply.

Reply:
I was referring to later with `data_len` but I see now what you are doing.